test/test_utils/ec2.py

```diff
@@ -487,26 +494,45 @@ def enforce_IMDSv2(instance_id, region=DEFAULT_REGION, ec2_client=None, hop_limi
     if not response:
         raise Exception("Unable to enforce IMDSv2. No response received.")
 
-    timeout = 3
     state = None
     if response["InstanceId"]:
-        while timeout > 0:
-            time.sleep(timeout)
-            LOGGER.info(f"Slept for: {timeout}")
-            res = ec2_client.describe_instances(InstanceIds=[instance_id])
-            if res:
-                metadata_options = res['Reservations'][0]['Instances'][0]['MetadataOptions']
-                http_tokens = metadata_options['HttpTokens']
-                state = metadata_options['State']
-                instance_hop_limit = metadata_options['HttpPutResponseHopLimit']
-
-                if http_tokens == 'required' and state == 'applied' and hop_limit == instance_hop_limit:
-                    break
-            timeout -= 1
-
-    if state == 'pending' or timeout == 0:
+        res = ec2_client.describe_instances(InstanceIds=[instance_id])
+        if res:
+            metadata_options = res['Reservations'][0]['Instances'][0]['MetadataOptions']
+            state = metadata_options['State']
+            LOGGER.info(f"Modify Metadata options of EC2 instance: {metadata_options}")
+    if state != "applied":
         raise Exception("Unable to enforce IMDSv2. Describe instance is not able to confirm if IMDSv2 enforced.")
-    LOGGER.info(f"Modify Metadata options State of EC2 instance: {state}")
+
+
+@retry(stop=stop_after_attempt(16), wait=wait_fixed(60))
+def enforce_IMDSv1(instance_id, region=DEFAULT_REGION, ec2_client=None):
+    """
+    Enabled IMDSv1 on EC2 instance.
+
+    :param instance_id: str, ec2 instance id
+    :param region: str, Region where ec2 instance is launched.
+    :param ec2_client: str, ec2 client.
+    :param hop_limit: str, hop limit to be set on ec2 instance.
+    """
+    ec2_client = ec2_client or get_ec2_client(region)
+    response = ec2_client.modify_instance_metadata_options(
+        InstanceId=instance_id,
+        HttpTokens='optional',
+        HttpPutResponseHopLimit=1
+    )
+
+    if not response:
+        raise Exception("Unable to enforce IMDSv1. No response received.")
+
+    if response["InstanceId"]:
+        res = ec2_client.describe_instances(InstanceIds=[instance_id])
+        if res:
+            metadata_options = res['Reservations'][0]['Instances'][0]['MetadataOptions']
+            state = metadata_options['State']
+            LOGGER.info(f"Modify Metadata options of EC2 instance: {metadata_options}")
+    if state != "applied":
```

Comment:
Not defining local variable 'state' before referencing can lead to UnboundLocalError.

Reply:
+1 to the comment from @kenny-ezirim.

This will technically raise an Exception in the correct circumstances, because something should have gone wrong for `state` to remain undefined until line 534. But the type of the exception would be wrong, and would confuse the dev reading the logs.

Reply:
Agree, this is copy paste code from another func to enfore imdv2, looks like I missed to add state=None